src/imitation/scripts/analyze.py

```diff
@@ -1,23 +1,74 @@
 from collections import OrderedDict
+import os
 import os.path as osp
-from typing import Optional
+import tempfile
+from typing import List, Optional
 
 import pandas as pd
 from sacred.observers import FileStorageObserver
+import tensorflow as tf
 
 from imitation.scripts.config.analyze import analysis_ex
 import imitation.util.sacred as sacred_util
 from imitation.util.sacred import dict_get_nested as get
 
 
-@analysis_ex.main
+@analysis_ex.command
+def gather_tb_directories(source_dir: str,
+                          run_name: Optional[str],
+                          env_name: Optional[str],
+                          skip_failed_runs: bool,
+                          ) -> dict:
+  """Gather Tensorboard directories from a `parallel_ex` run.
+  The directories are copied to a unique directory in `/tmp/analysis_tb/` under
+  subdirectories matching the Tensorboard events' Ray Tune trial names.
+
+  Undocumented arguments are the same as in `analyze_imitation()`.
+
+  Args:
+    source_dir: A local_dir for Ray. For example, `~/ray_results/`.
+
+  Returns:
+    A dict with two keys. "gather_dir" (str) is a path to a /tmp/
+    directory containing all the TensorBoard runs filtered from `source_dir`.
+    "n_tb_dirs" (int) is the number of TensorBoard directories that were
+    filtered.
+  """
+  sacred_dicts = _get_sacred_dicts(
+    source_dir, run_name, env_name, skip_failed_runs)
+  os.makedirs("/tmp/analysis_tb", exist_ok=True)
+  tmp_dir = tempfile.mkdtemp(dir="/tmp/analysis_tb/")
+
+  tb_dirs_count = 0
+  for basename in ["tb", "sb_tb"]:
+    os.mkdir(osp.join(tmp_dir, basename))  # Build symlinks in the directory.
+    for sd in sacred_dicts:
+      # Expecting a path like "~/ray_results/{run_name}/sacred/1".
+      # Want to search for all Tensorboard dirs inside
+      # "~/ray_results/{run_name}".
+      sacred_dir = sd.sacred_dir.rstrip("/")
+      run_dir = osp.dirname(osp.dirname(sacred_dir))
+      run_name = osp.basename(run_dir)
+      for tb_src_dir in sacred_util.filter_subdirs(
+          run_dir, lambda path: osp.basename(path) == basename):  # noqa: E125
+        tb_dest_dir = osp.join(tmp_dir, basename, run_name)
+        os.symlink(tb_src_dir, tb_dest_dir)
```

Comment:
`tb_dest_dir`, not `tb_src_dir`, is the new symlink file needs to be unique. (Following the same src/dst naming convention as https://docs.python.org/3/library/os.html#os.symlink )

Reply:
I renamed `tb_dest_dir` to `tb_symlink` to make this less ambiguous.

Reply:
I still don't get this. `os.symlink(tb_src_dist, tb_dest_dir)` creates a symlink from `tb_dest_dir` to `tb_src_dir`. So if there are multiple `tb_src_dir`, then `tb_dest_dir` will point only to the last `tb_src_dir`, not to any of the previous ones.

```
>>> os.symlink('foo', 'bar')
>>> os.symlink('fiz', 'bar')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 17] File exists
>>> os.system('ls -la bar')
lrwxrwxrwx 1 adam adam 3 Nov  6 16:10 bar -> foo
0
```

Reply:
I think the code works OK since you never have more than one TensorBoard log directory. But this feels like something that should be an explicit assertion.

Reply:
That's right, I've been implicitly assuming only 1 of each type of TB log directory

Reply:
That's probably a reasonable assumption, but is not what you normally use a `for` loop for! How about something like:

```
tb_src_dirs = list(filter_subdirs(...))
assert len(tb_src_dirs) <= 1, "should be at most one TB dir of each type"
if tb_src_dirs:  # this could be OK as a for loop too but a bit confusing IMO
  tb_src_dir = tb_src_dirs[0]
  ...
```